File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> left_product(n);
        vector<int> right_product(n);
        vector<int> ans(n);

        left_product[0] = nums[0];
        right_product[n-1] = nums[n-1];

        for(int i=1; i < n ; i++) {
            left_product[i] =  left_product[i-1] * nums[i];
            right_product[n-i-1] = right_product[n-i] * nums[n-i-1];
        }

        ans[0] = right_product[1];
        ans[n-1] = left_product[n-2];

        for(int i=1; i<n-1; i++) {
            ans[i] = left_product[i-1] * right_product[i+1];
        }
        return ans;
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.java
class Solution {
    private Set<List<Integer>> ans = new HashSet<>();
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        addToAns(new ArrayList<>(), candidates, 0, target, 0);
        return ans.stream().collect(Collectors.toList());
    }
    public void addToAns(ArrayList<Integer> nums, int[] candidates, int i, int target, int sum) {
        if(sum==target) {
            // System.out.println("YESSSSSSS");
            ans.add(new ArrayList<>(nums));
        }
        // System.out.println(nums);
        // System.out.println(i + " "  + sum + " " + target);
        
        for(;i<candidates.length;i++) {
            nums.add(candidates[i]);
            sum+=candidates[i];

            addToAns(nums, candidates, i+1, target, sum);
            nums.remove(nums.size()-1);
            sum-=candidates[i];
        }
        
    }
}

class Solution_Previous_Question {
    private List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        ans.add(new ArrayList<>());
        addSubsets(new ArrayList<>(), nums, 0);
        return ans;
    }
    public void addSubsets(ArrayList<Integer> already, int[] nums, int i) {
        for(;i<nums.length;i++) {
            List<Integer> temp = new ArrayList<>(already);
            temp.add(nums[i]);
            ans.add(temp);
            // ans.add((new ArrayList<>(already)).add(nums[i]));
            already.add(nums[i]);
            addSubsets(already, nums, i+1);
            already.remove(already.size()-1);
        }
    }
}